File: sorting_algos/bubble_sort/bubble_sort.c
/*
    Bubble sort

    set a swap counter to a non-zero value
    until the swap counter is 0 (meaning no swaps were performed):
        set the swap counter to 0
        Iterate over array, comparing each pair (where each pair is an element and the element after it) 
            and switching them if they are out of order
        If any switches are performed, increment the swap counter
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void createRandArray(int arr[], int size, int maxVal);
void printArray(int arr[], int size);
void bubbleSort(int arr[], int size);

int main(void){

    // seed the random generator
    srand((unsigned)time(NULL));

    int size, maxValue;

    printf("Size of array: ");
    scanf("%d", &size);
    printf("Maximum value: ");
    scanf("%d", &maxValue);

    // create array
    int array[size];
    createRandArray(array, size, maxValue);

    printf("Original array: ");
    printArray(array, size);

    // sort array
    bubbleSort(array, size);

    printf("Sorted array: ");
    printArray(array, size);

}

void createRandArray(int arr[], int size, int maxVal){
    
    for (int i = 0; i < size; i++){
        arr[i] = rand() % (maxVal + 1);
    }

}

void printArray(int arr[], int size){

    for (int i = 0; i < size; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");

}

void bubbleSort(int arr[], int size){

    int swap_counter = -1, tempVal;

    while (swap_counter != 0){
        swap_counter = 0;

        for (int i = 0; i < size - 1; i++){
            tempVal = arr[i+1];
            if (arr[i] > arr[i+1]){
                arr[i+1] = arr[i];
                arr[i] = tempVal;
                swap_counter++;
            }
        }
    }
}

File: sorting_algos/bubble_sort/bubble_sort.py
from random import randint

def create_array(size, maxVal):
    arr = []
    for i in range(0, size):
        arr.append(randint(0, maxVal))
    return arr

# print array elements with each element separated with a space, then print new line
def print_array(arr):
    for i in range(0, len(arr)):
        print(str(arr[i]), end = " ")
    print()

def bubble_sort(arr):
    swap_counter = -1

    while (swap_counter != 0):
        swap_counter = 0

        for i in range(0, len(arr)-1):
            if (arr[i] > arr[i + 1]):
                temp_val = arr[i+1]
                arr[i+1] = arr[i]
                arr[i] = temp_val
                swap_counter += 1
    
    return arr
                

def main():
    # for now let the array size be 50 and max integer be 100
    size = 10
    maxVal = 100

    array = create_array(size, maxVal)

    print_array(array)

    # now sort the array
    array = bubble_sort(array)

    print_array(array)



if __name__ == "__main__":
    main()